File: wrapper/lld_wrapper.py
import os
import sys
import subprocess
from sys import exit
import toolchain_wrapper_tools

DIR=os.path.dirname(__file__)

def main(target, exec_name):
    args = sys.argv[1:]
    lld_name = 'ld.lld'
    if 'apple' in target:
        lld_name = 'ld64.lld'
    elif 'wasm' in target:
        lld_name = 'wasm-ld'
    elif 'msvc' in target:
        if len(args) > 0 and args[0].lower() == '/lib':
            # Generate static lib, act as lib.exe. It must be the first argument
            lld_name = 'lld-link'
        else:
            # Invoke lld-link via clang frontend to get MSVC lib path
            lld_name = '%s-clang' % target
            args = ['/clang:-Wl,' + arg for arg in args ]
    
    exec_prog = os.path.join(DIR, lld_name)
    toolchain_wrapper_tools.exec_subprocess([exec_prog] + args)

File: wrapper/lld_wrapper.go
package main

import (
	"path/filepath"
	"strings"
)

func lldWrapperMain(execDir string, target string, execName string, cmdArgv []string) {
	cmdArgv = cmdArgv[1:]
	lldName := "ld.lld"
	if strings.Contains(target, "apple-") {
		lldName = "ld64.lld"
	} else if strings.HasPrefix(target, "wasm") {
		lldName = "wasm-ld"
	} else if strings.Contains(target, "msvc") {
		if len(cmdArgv) > 0 && strings.ToLower(cmdArgv[0]) == "/lib" {
			// Generate static lib, act as lib.exe. It must be the first argument
			lldName = "lld-link"
		} else {
			// Invoke lld-link via clang frontend to get MSVC lib path
			lldName = target + "-clang"
			for i, arg := range cmdArgv {
				cmdArgv[i] = "/clang:-Wl," + arg
			}
		}
	}
	lldExec := filepath.Join(execDir, lldName)
	runCommand(lldExec, cmdArgv, nil)
}
